File: csharp/Minesweeper.Tests/E2ETests/MinesweeperTests.cs
using MineSweeper;
using MineSweeper.Engine;

namespace MineSweeper.Tests.E2ETests
{
    [TestFixture]
    public class MinesweeperTests
    {
        [Test]
        public void TestGameFlow_WinScenario_AllNonMineCellsSelected()
        {
            // Assuming there's a setup or constructor that initializes the game
            var game = new GameConsole();
            game.Initialize(4, [new(0, 0), new(1, 1), new(2, 2), new(3, 3)]);

            // Simulate selecting all non-mine cells
            var nonMineCells = new Point[] { new(0, 1), new(0, 2), new(0, 3), new(1, 0), new(1, 2), new(1, 3), new(2, 0), new(2, 1), new(2, 3), new(3, 0), new(3, 1), new(3, 2) };
            foreach (var cell in nonMineCells)
            {
                game.SelectCell(cell);
            }

            // Assert that the game is won
            Assert.That(game.IsGameWon(), "The game should be won when all non-mine cells are uncovered.");
        }


        [Test]
        public void TestGameFlow_WinScenario_RevealingAdjacentCells()
        {
            // Setup game with predetermined mine locations for repeatability
            var game = new GameConsole();
            game.Initialize(4, [new(3, 3)]);

            // Simulate user interactions
            game.SelectCell(0, 0); // Safe, will reveal adjacent cells with no adjacent mines
            game.SelectCell(2, 3); // Safe
            game.SelectCell(2, 2); // Safe
            game.SelectCell(3, 2); // Safe

            Assert.Multiple(() =>
            {
                // Assert that the adjacent cells are correctly revealed
                Assert.That(game.Grid.GetCell(0, 1).IsRevealed, "The adjacent cell (0, 1) should be revealed.");
                Assert.That(game.Grid.GetCell(1, 0).IsRevealed, "The adjacent cell (1, 0) should be revealed.");
                Assert.That(game.Grid.GetCell(1, 1).IsRevealed, "The adjacent cell (1, 1) should be revealed.");

                // Assert the game is won
                Assert.That(game.IsGameWon());
            });
        }

        [Test]
        public void TestGameFlow_LossScenario()
        {
            // Set up the game with known mine positions
            var game = new GameConsole();
            game.Initialize(4, [new(0, 1), new(1, 3)]);

            // Act by selecting the cell with a mine
            game.SelectCell(0, 1);

            // Assert that the game is over
            Assert.That(game.IsGameLost(), "The game should be over after a mine is uncovered.");
        }

        [Test]
        public void TestGameFlow_InvalidCellSelection()
        {
            // Set up the game with known mine positions
            var game = new GameConsole();
            game.Initialize(4, [new(1, 1)]);

            // Act by selecting an invalid cell & Assert that an exception is thrown
            Assert.Throws<ArgumentOutOfRangeException>(() => game.SelectCell(5, 5), "An invalid cell selection should throw an exception.");
        }

        [Test]
        public void TestGameFlow_RevealMineCell()
        {
            // Set up the game with known mine positions
            var game = new GameConsole();
            game.Initialize(4, [new(2, 2)]);

            // Act by selecting a mine cell
            game.SelectCell(2, 2);

            // Assert that the game is lost
            Assert.That(game.IsGameLost(), "The game should be lost after selecting a mine cell.");
        }

        [Test]
        public void TestGameFlow_UninitializedGame()
        {
            var game = new GameConsole();
            Assert.Throws<InvalidOperationException>(() => game.Start(), "An uninitialized game should throw an exception when started.");
        }
    }
}

File: kotlin/src/test/kotlin/e2etests/MineSweeperTests.kt
package e2etests

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.minesweeper.GameConsole
import org.minesweeper.engine.Point

internal class MineSweeperTests {

    @Test
    fun `Test game flow - Win scenario all non-mine cells selected`() {
        val game = GameConsole()
        game.initialize(4, listOf(Point(0, 0), Point(1, 1), Point(2, 2), Point(3, 3)))
        val nonMineCells = listOf(Point(0, 1), Point(0, 2), Point(0, 3), Point(1, 0), Point(1, 2), Point(1, 3), Point(2, 0), Point(2, 1), Point(2, 3), Point(3, 0), Point(3, 1), Point(3, 2))
        nonMineCells.forEach { game.selectCell(it) }
        assertTrue(game.isGameWon(), "The game should be won when all non-mine cells are uncovered.")
    }

    @Test
    fun `Test game flow - Win scenario revealing adjacent cells`() {
        val game = GameConsole()
        game.initialize(4, listOf(Point(3, 3)))
        game.selectCell(0, 0)  // Safe, will reveal adjacent cells with no adjacent mines

        assertAll("Check if adjacent cells are revealed",
            { assertTrue(game.grid.getCell(0, 1).isRevealed) },
            { assertTrue(game.grid.getCell(1, 0).isRevealed) },
            { assertTrue(game.grid.getCell(1, 1).isRevealed) },
            { assertTrue(game.isGameWon()) }
        )
    }

    @Test
    fun `Test game flow - Loss scenario`() {
        val game = GameConsole()
        game.initialize(4, listOf(Point(0, 1), Point(1, 3)))
        game.selectCell(0, 1)
        assertTrue(game.isGameLost(), "The game should be over after a mine is uncovered.")
    }

    @Test
    fun `Test game flow - Invalid cell selection`() {
        val game = GameConsole()
        game.initialize(4, listOf(Point(1, 1)))
        assertThrows<IndexOutOfBoundsException> { game.selectCell(5, 5) }
    }

    @Test
    fun `Test game flow - Reveal mine cell`() {
        val game = GameConsole()
        game.initialize(4, listOf(Point(2, 2)))
        game.selectCell(2, 2)
        assertTrue(game.isGameLost(), "The game should be lost after selecting a mine cell.")
    }

    @Test
    fun `Test game flow - Uninitialized game`() {
        val game = GameConsole()
        assertThrows<IllegalStateException> { game.start() }
    }
}
